### PR TITLE
Add consumer pause & resume

### PR DESCRIPTION
**Pull request by Nevon**
This also fixes a bug where if you specify `0` as the partition for a message, the partitioner would fall back to using the hash of the key, as it would consider `message.partition` falsy.

### CODE REVIEW

**Comment by soroushhakami** [comment]
👏

**Comment by tulios** [comment]
Looks good, I would just mention in the readme that the topic won't be fetched in the next cycle, you might get messages from the paused topic in the current batch.

**Comment by tulios** [comment]
You can also add the signature of the methods in the readme and mention that the library currently doesn't support specific partitions.

**Comment by Nevon** [comment]
This works fine as long as you are subscribed to multiple topics, and don't pause all of them. If you do, however, it looks like the fetch loop becomes synchronous, so nothing else ever gets a chance to execute.

**Comment by Nevon** [comment]
@tulios: 🏖💻🍹

File: src/consumer/consumerGroup.js [hunk 1]
```diff
@@ -160,6 +162,24 @@ module.exports = class ConsumerGroup {
...
+  }
+
+  resume(topics) {
+    this.logger.info(`Resuming fetching from ${topics.length} topics`, {
+      topics,
+    })
+    this.subscriptionState.resume(topics)
+  }
+
+  paused() {
```
Comment (joaomilho) [inline src/consumer/consumerGroup.js]: small stuff, but wouldn't a `get paused()` be a nicer API?

File: src/consumer/subscriptionState.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+module.exports = class SubscriptionState {
+  constructor() {
+    this.pausedTopics = new Set()
```
Comment (joaomilho) [inline src/consumer/subscriptionState.js]: fancy

File: src/producer/partitioners/default/index.js [hunk 3]
```diff
@@ -26,7 +26,7 @@ module.exports = () => {
     const availablePartitions = partitionMetadata.filter(p => p.leader >= 0)
     const numAvailablePartitions = availablePartitions.length
 
-    if (message.partition) {
+    if (message.partition !== null && message.partition !== undefined) {
```
Comment (joaomilho) [inline src/producer/partitioners/default/index.js]: why? Is `false` a valid partition?

Comment (Nevon) [reply src/producer/partitioners/default/index.js]: No, but `0` is.

File: src/consumer/subscriptionState.js (hunk 2 above)
Comment (Nevon) [reply src/consumer/subscriptionState.js]: Don't want any of those filthy duplicates in here. 💅

**Comment by Nevon** [comment]
This is as fascinating as it is terrifying. I'm running my consumer and pausing/resuming by calling an HTTP endpoint. If I pause consuming of all topics, KafkaJS goes into a busy loop that won't let me call the endpoint to resume consuming.

The culprit appears to be: https://github.com/tulios/kafkajs/blob/b474e53ad5d08fdab216702c57b1bf013edd70d8/src/consumer/consumerGroup.js#L265 in `ConsumerGroup#fetch`. `requests` will be an empty array, so my thinking was that maybe `await Promise.all([])` resolves synchronously or something (doesn't make any sense, but nothing makes sense in this world). So I tried changing it to:

```javascript
const results = await Promise.all(requests.length > 0 ? requests : [new Promise(resolve => process.nextTick(() => resolve([])))])
```

This did not help. At this point I _know_ that it resolves asynchronously, but still, express never got a chance to do anything.

So, let's try with `setTimeout`:

```javascript
const results = await Promise.all(requests.length > 0 ? requests : [new Promise(resolve => setTimeout(() => resolve([]), 0))])
```

This works.

At this point someone needs to explain to me how the event loop works, because clearly it does not work the way I think it does.

**Comment by Nevon** [comment]
I'm still confused, but at least I now know why `process.nextTick` didn't work. https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/#process-nexttick

Scheduling fetch using `setImmediate` makes more sense.

**Comment by Nevon** [comment]
I've been running this for a while now, and it seems to be working fine. However, if you're running with a loglevel >= info, you'll produce an insane amount of logs, as each fetch loop logs, and without any topics to fetch from, it will basically produce logs as fast as it can.

I think it makes sense to add some throttling to the fetch loop, at least when you're not subscribed to any topic.

**Pull request merged**